server/forge/bitbucketserver/bitbucketserver.go

```diff
@@ -0,0 +1,625 @@
+// Copyright 2022 Woodpecker Authors
+// Copyright 2018 Drone.IO Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bitbucketserver
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/mrjones/oauth"
+	bb "github.com/neticdk/go-bitbucket/bitbucket"
+	"github.com/rs/zerolog/log"
+
+	"github.com/woodpecker-ci/woodpecker/server/forge"
+	"github.com/woodpecker-ci/woodpecker/server/forge/bitbucketserver/internal"
+	"github.com/woodpecker-ci/woodpecker/server/forge/common"
+	forge_types "github.com/woodpecker-ci/woodpecker/server/forge/types"
+	"github.com/woodpecker-ci/woodpecker/server/model"
+	"github.com/woodpecker-ci/woodpecker/server/store"
+)
+
+const (
+	requestTokenURL   = "%s/plugins/servlet/oauth/request-token"
+	authorizeTokenURL = "%s/plugins/servlet/oauth/authorize"
+	accessTokenURL    = "%s/plugins/servlet/oauth/access-token"
+
+	secret = "045dfb11b042c3c44d68274fd22338e0" // TODO: Temporary
```

Comment:
Should this secret be hardcoded here or is it a leftover?

Reply:
Good catch - it _is_ a left-over - kind of.

There is an "extra" security feature in the Bitbucket webhook such that when you register the webhook with Bitbucket you also provide a symmetric-key. Then when Bitbucket invokes the webhook it generates a signature using the symmetric-key such that the receiver (Woodpecker in this case) can validate that it was in fact a trusted call.

It would be no problem to generate the HMAC key however the `Hook` function only gets the `context.Context` and the `http.Request` - and before validating the incoming request that is pretty much all there is.

I see that there is no similar validation in the other Git integrations. Is there anywhere to store such a thing? Basically I guess it should be generated once and probably stored somewhere in the database.

Reply:
One option could be to provide it through the configuration, of course, like the keys used to register Woodpecker as "application" in Bitbucket 🤔

Reply:
Actually it seems that the jwt token part of the webhook url is signed using the "repo_hash" field from the db as HMAC key. In theory it should be possible to use the repo_hash also be for the Btibucket webhook signing feature.

Problem is that the repo (or remote forge id) is part of the signed payload and it is needed to fetch the key. So to fetch the signing key it would be required to parse the payload and afterwards validated - it is a little bit backwards 🤔

Reply:
I have changed the code so that the `repo.Hash` is used as HMAC key for both the jwt in the webhook url (as it is for all webhooks) and then for the Bitbucket feature of signing the webhook payload.